src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/DiamondOperatorForVariableDefinitionCheck.java

```diff
@@ -0,0 +1,122 @@
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.checks.coding;
+
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+
+/**
+ * This Check highlights variable definition statements where <a href=
+ * "http://docs.oracle.com/javase/7/docs/technotes/guides/language/type-inference-generic-instance-creation.html">
+ * diamond operator</a> could be used.<br>
+ * <b>Rationale</b>: using diamond operator (introduced in Java 1.7) leads to shorter code<br>
+ * and better code readability. It is suggested by Oracle that the diamond primarily using<br>
+ * for variable declarations.<br><br>
+ * E.g. of statements:
+ *
+ * <p><b>Without diamond operator:</b><br>
+ * {@code
+ * Map&lt;String, Map&lt;String, Integer&gt;&gt; someMap =
+ *     new HashMap&lt;String, Map&lt;String, Integer&gt;&gt;();
+ * }<br>
+ * <b>With diamond operator:</b><br>
+ * {@code
+ * Map&lt;String, Map&lt;String, Integer&gt;&gt; someMap = new HashMap&lt;&gt;();
+ * }</p>
+ *
+ * @author <a href="mailto:nesterenko-aleksey@list.ru">Aleksey Nesterenko</a>
+ */
+public class DiamondOperatorForVariableDefinitionCheck extends AbstractCheck {
+
+    /**
+     * A key is pointing to the warning message text in "messages.properties"
+     * file.
+     */
+    public static final String MSG_KEY = "diamond.operator.for.variable.definition";
+
+    @Override
+    public int[] getDefaultTokens() {
+        return new int[] {TokenTypes.VARIABLE_DEF};
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getDefaultTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return getDefaultTokens();
+    }
+
+    @Override
+    public void visitToken(DetailAST variableDefNode) {
+
+        final DetailAST assignNode = variableDefNode.findFirstToken(TokenTypes.ASSIGN);
+
+        // we check only creation by NEW
+        if (assignNode != null && assignNode.getFirstChild().getFirstChild().getType()
+                == TokenTypes.LITERAL_NEW) {
+
+            final DetailAST newNode = assignNode.getFirstChild().getFirstChild();
+
+            final DetailAST variableDefNodeType =
+                        variableDefNode.findFirstToken(TokenTypes.TYPE);
+            final DetailAST varDefArguments = getFirstTypeArgumentsToken(variableDefNodeType);
+
+            // generics has to be on left side
+            if (varDefArguments != null
+                && newNode.getLastChild().getType() != TokenTypes.OBJBLOCK
+                // arrays can not be generics
+                && newNode.findFirstToken(TokenTypes.ARRAY_DECLARATOR) == null) {
+
+                final DetailAST typeArgs = getFirstTypeArgumentsToken(newNode);
+
+                if (varDefArguments.equalsTree(typeArgs)) {
```

Comment:
> override the implementation in DetailAST?

We tried overriding `equals` method in that class and it lead to failures everywhere else.
I don't know what will happen if we try to override this method to.
I would just keep implementation in this class.